File: Test Driven Development/sources/tdd-bowling-kata/java/src/main/java/kata/BowlingGame.java
package kata;

public class BowlingGame {

    private int rolls[] = new int[21];
    private int currentRoll = 0;

    public void roll(int pins) {
        rolls[currentRoll++] = pins;
    }

    public int score() {
        int score = 0;
        int frameIndex = 0;
        for (int frames = 0; frames < 10; frames++) {
            if (isSpire(frameIndex)) {
                score += 10 + rolls[frameIndex + 2];
                frameIndex += 2;
            } else {
                score += rolls[frameIndex] + rolls[frameIndex + 1];
                frameIndex += 2;
            }
        }
        return score;
    }

    private boolean isSpire(int i) {
        return rolls[i] + rolls[i + 1] == 10;
    }
}

File: Test Driven Development/sources/tdd-bowling-kata/objective-c/BowlingGameTests/BowlingGame.m
#import "BowlingGame.h"

@implementation BowlingGame

- (int)score {
    int score = 0;
    int ballIndex = 0;
    for (int frame = 0; frame < 10; frame++) {
        if ([self isStrike:ballIndex]) {
            score += 10 + [self strikeBonus:ballIndex];
            ballIndex++;
        }
        else if ([self isSpare:ballIndex]) {
            score += 10 + [self spareBonus:ballIndex];
            ballIndex += 2;
        }
        else {
            score += [self sumOfBalls:ballIndex];
            ballIndex += 2;
        }
    }
    return score;
}

- (int)sumOfBalls:(int)ballIndex {
    return rolls[ballIndex] + rolls[ballIndex + 1];
}

- (int)spareBonus:(int)ballIndex {
    return rolls[ballIndex + 2];
}

- (int)strikeBonus:(int)ballIndex {
    return rolls[ballIndex + 1] + rolls[ballIndex + 2];
}

- (BOOL)isStrike:(int)ballIndex {
    return rolls[ballIndex] == 10;
}

- (BOOL)isSpare:(int)currentFrame {
    return rolls[currentFrame] + rolls[currentFrame + 1] == 10;
}

- (void)roll:(int)pins {
    rolls[currentRoll++] = pins;
}

@end